pykafka/cli/kafka_tools.py

```diff
@@ -0,0 +1,285 @@
+import argparse
+import calendar
+import datetime as dt
+
+import tabulate
+
+import pykafka
+from pykafka.common import OffsetType
+
+#
+# Helper Functions
+#
+
+
+def fetch_offsets(client, topic, offset):
+    """Fetch raw offset data from a topic.
+
+    :param client: KafkaClient connected to the cluster.
+    :type client:  :class:`pykafka.KafkaClient`
+    :param topic:  Name of the topic.
+    :type topic:  :class:`pykafka.topic.Topic`
+    :param offset: Offset to reset to. Can be earliest, latest or a datetime.
+        Using a datetime will reset the offset to the latest message published
+        *before* the datetime.
+    :type offset: :class:`pykafka.common.OffsetType` or
+        :class:`datetime.datetime`
+    :returns: dict of {partition_id: :class:`pykafka.protocol.OffsetPartitionResponse`}
+    """
+    if offset.lower() == 'earliest':
+        return topic.earliest_available_offsets()
+    elif offset.lower() == 'latest':
+        return topic.latest_available_offsets()
+    else:
+        offset = dt.datetime.strptime(offset, "%Y-%m-%dT%H:%M:%S")
+        offset = int(calendar.timegm(offset.utctimetuple())*1000)
+        return topic.fetch_offset_limits(offset)
```

Comment:
Yeah, it's crazy. It's based on segment boundaries in the data files.  The very end of [this post](http://cfchou.github.io/blog/2015/04/23/a-closer-look-at-kafka-offsetrequest/) sort of explains it.